CryptomatorFileProvider/Workflow/WorkflowDependencyFactory.swift

```diff
@@ -0,0 +1,147 @@
+import CryptomatorCloudAccessCore
+import Foundation
+import Promises
+
+/**
+ Workflow Dependency Factory
+
+ The workflow dependency factory holds a graph consisting of lock and unlock nodes, where a node can be of type read or write, for currently running/pending workflows.
+
+ This makes it possible to create dependencies between individual workflows. This prevents, for example, that a file is uploaded to path `/a/b` before the corresponding folder is created on path `/a` - provided that the two actions have been executed locally in the correct order.
+ */
+class WorkflowDependencyFactory {
+	private let readTasks: WorkflowDependencyTaskCollectionType
+	private let writeTasks: WorkflowDependencyTaskCollectionType
+
+	init(readTasks: WorkflowDependencyTaskCollectionType, writeTasks: WorkflowDependencyTaskCollectionType) {
+		self.readTasks = readTasks
+		self.writeTasks = writeTasks
+	}
+
+	convenience init() {
+		self.init(readTasks: WorkflowDependencyTaskCollection(), writeTasks: WorkflowDependencyTaskCollection())
+	}
+
+	/**
+	 Creates dependencies for the passed paths with the passed lock type.
+
+	 - Parameter paths: The paths for which dependencies are to be created. These are the full paths that affect a workflow.
+	 - Parameter lockType: The lock type used to construct the leaf nodes (the nodes for the respective full path).
+	 */
+	public func createDependencies(paths: [CloudPath], lockType: LockType) -> WorkflowDependency {
+		let unlock = Promise<Void>.pending()
+		var locks = [Promise<Void>]()
+		paths.forEach { path in
+			let lock = createLock(path: path, type: lockType)
+			locks.append(lock)
+			createUnlock(path: path, type: lockType, basedOn: unlock)
+		}
+		let lock: Promise<Void> = all(locks).then { _ -> Void in
+			// no-op
+		}
+		return WorkflowDependency(lock: lock, unlock: unlock)
+	}
+
+	/**
+	 Creates dependencies for the passed path with the passed lock type.
+
+	 - Parameter path: The path for which dependencies are to be created. These is the full path that affect a workflow.
+	 - Parameter lockType: The lock type used to construct the leaf nodes (the nodes for the respective full path).
+	 */
+	public func createDependencies(for path: CloudPath, lockType: LockType) -> WorkflowDependency {
+		return createDependencies(paths: [path], lockType: lockType)
+	}
+
+	/**
+	 Creates chain of "lock" promises for the given path.
+
+	 For the passed path there is a top-down dependency between the promises, i.e. a promise for the child path depends on the promise for the parent path.
+	 */
+	private func createLock(path: CloudPath, type: LockType) -> Promise<Void> {
+		var dependencies = [Promise<Void>]()
+		if let parentPath = path.getParent() {
+			let parentLock = createLock(path: parentPath, type: .read)
+			dependencies.append(parentLock)
+		}
+		// All tasks wait for other write tasks on the same path
+		let writeTasksForPath = writeTasks[path]
+		dependencies.append(contentsOf: writeTasksForPath)
```

Comment:
:warning: is it correct to wait for `all` dependencies? I think we must wait for "fulfilled" parent and "fulfilled or rejected" other dependencies.

```suggestion
		// All tasks wait for other write tasks on the same path
		let writeTasksForPath = writeTasks[path]
		dependencies.append(fulfilledOrRejected(writeTasksForPath))
```